### PR TITLE
Developer Tutorial Move - 2 (Command Local)

### PR DESCRIPTION
**Pull request by kb1lqd**
# Pull Request Overview

This pull request updates the `Commanding-Local` tutorial script and readme.md.

## Notable Updates
* Spellchecking and other typo's fixed is both readme and example script
* General wording + partial PEP 8 update (easy stuff)
* Updated script to use an INI file for configuration
* Added exception handling to ECHO command for both good practice and due to noted bug ( #50 ) 
  * Added section in readme about the workaround

## Testing Notes

* I've verified it works as written in readme and intended by script

### CODE REVIEW

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py [hunk 1]
```diff
@@ -1,87 +1,110 @@
...
 
-#Variables
-local_device_callsign = 'REPLACEME'  # Should match the connected Faraday unit as assigned in Proxy configuration
-local_device_node_id = REPLACEME  # Should match the connected Faraday unit as assigned in Proxy configuration
+# Variables
+local_device_callsign = config.get("DEVICES",
+                                   "UNIT0CALL")  # Should match the connected Faraday unit as assigned in Proxy configuration
+local_device_node_id = config.get("DEVICES",
+                                  "UNIT0ID")  # Should match the connected Faraday unit as assigned in Proxy configuration
+local_device_callsign = str(local_device_callsign).upper()
```
Comment (kb1lqc) [inline Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py]: config.get() is already a string and just perform .upper() to it on the same line.

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py [hunk 2]
```diff
@@ -1,87 +1,110 @@
...
+faraday_1 = faradaybasicproxyio.proxyio()  # default proxy port
 faraday_cmd = faradaycommands.faraday_commands()
 
-
 ##############
 ## TOGGLE GPIO
 ##############
-#Turn LED 1 ON LOCAL
+# Turn LED 1 ON LOCAL
 print "Turning ON the Green LED (LED #1)"
```
Comment (kb1lqc) [inline Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py]: I won't block on it but we should really use logging instead of print.

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py [hunk 3]
```diff
@@ -1,87 +1,110 @@
...
 
+# Definitions
 
-#Variables
-local_device_callsign = 'REPLACEME'  # Should match the connected Faraday unit as assigned in Proxy configuration
-local_device_node_id = REPLACEME  # Should match the connected Faraday unit as assigned in Proxy configuration
+# Variables
+local_device_callsign = config.get("DEVICES",
+                                   "UNIT0CALL")  # Should match the connected Faraday unit as assigned in Proxy configuration
+local_device_node_id = config.get("DEVICES",
```
Comment (kb1lqc) [inline Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py]: might want to use .getint()

https://docs.python.org/2/library/configparser.html#ConfigParser.RawConfigParser.getint

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py [hunk 4]
```diff
@@ -1,87 +1,110 @@
...
 print "Turning ON the Red LED (LED #2)"
 command = faraday_cmd.CommandLocalGPIOLED2On()
 faraday_1.POST(local_device_callsign, local_device_node_id, faraday_1.CMD_UART_PORT, command)
 time.sleep(1)
 
-#Turn LED 2 OFF LOCAL
+# Turn LED 2 OFF LOCAL
 print "Turning Off the Red LED (LED #2)"
-command = faraday_cmd.CommandLocalGPIO(0, 0, 0, gpioallocations.LED_2, 0, 0) #This examples how the non predefined LED GPIO commanding is created. Multiple GPIO's can be toggled at once using ||'s
+command = faraday_cmd.CommandLocalGPIO(0, 0, 0, gpioallocations.LED_2, 0, 0)  # This examples how the non predefined LED GPIO commanding is created.
```
Comment (kb1lqc) [inline Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py]: wait why not use `CommandLocalGPIOLED2Off()?` like you used for LED1 since you used a similar command for LED1? Based on inline comment that's the goal?

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py [hunk 5]
```diff
@@ -1,87 +1,110 @@
...
+rx_echo_raw = faraday_1.GETWait(local_device_callsign, local_device_node_id, faraday_1.CMD_UART_PORT,
+                                sec_timeout=3)  # Wait for up to 3 seconds for data to arrive
+
+try:
+    # Decode BASE 64 encoded data from Proxy - Only using the [0] index packet (first) if multiple are retrieved
+    b64_data = rx_echo_raw[0]['data']
+    echo_decoded = faraday_1.DecodeRawPacket(b64_data)
+
+    # Display information
+    print "**Sending**\n"
```
Comment (kb1lqc) [inline Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py]: I feel like this should print prior to actually sending the data

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py [hunk 6]
```diff
@@ -1,87 +1,110 @@
...
+    print "Original Message: ", originalmsg
+    print "\n**Receiving**\n"
+    print "Decoded received ECHO'd Message:", echo_decoded  # Note that ECHO sends back a fixed packed regardless. Should update to send back exact length.
+    print "\nRAW Received BASE64 ECHO'd Message:", b64_data
+    print "\nDecoded BASE64 RAW Bytes:", repr(echo_decoded)
+
+except Exception as e:
```
Comment (kb1lqc) [inline Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py]: Broad exceptions like this aren't very good. Can let this one pass I guess but we should also try/except on specific lines we know what will fail and how it will fail. Those print statements are unlikely to cause problems we would not already have caught elsewhere.

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/configuration-template.ini [hunk 7]
```diff
@@ -0,0 +1,4 @@
+[DEVICES]
+UNITS=1
+UNIT0CALL=REPLACEME
+UNIT0ID= REPLACEME
```
Comment (kb1lqc) [inline Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/configuration-template.ini]: Remove space after `=`

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md [hunk 8]
```diff
@@ -1,29 +1,33 @@
...
+* Properly configured and connected proxy
+  * 1x Faraday
 
-## Start The Proxy Interface
+# Running The Tutorial Example Script
 
-Following the [Configuring Proxy](../../0-Welcome_To_Faraday/Configuring_Proxy/) tutorial configure, start, and ensure a successful connection to a locally (USB) connected Faraday digital radio.
+## Configuration
 
+* Open `configuration-template.ini` with a text editor
```
Comment (kb1lqc) [inline Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md]: I would use the `xxxxx.sample.ini` convention proposed in #110

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md [hunk 9]
```diff
@@ -1,29 +1,33 @@
...
+* Update `CALLSIGN` Replace `REPLACEME` to match the callsign of the Faraday unit **as assigned** in proxy
```
Comment (kb1lqc) [inline Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md]: Funky wording
>Update `CALLSIGN` Replace `REPLACEME` to match

Update `CALLSIGN` from `REPLACEME` to match the callsign....

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md [hunk 10]
```diff
@@ -1,29 +1,33 @@
...
+* Update `NODEID` to match the callsign node ID of the Faraday unit **as assigned** in proxy
```
Comment (kb1lqc) [inline Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md]: Probably the same as `CALLSIGN` as I suggested above

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md [hunk 11]
```diff
@@ -1,29 +1,33 @@
...
+* Save the file as `configuration.ini`
 
-## Tutorial Output Examples
 
-below is a screen-shot of the partial output of the tutorial script when run in a python interpreter (PyCharm). Be sure to look at the two LED's on Faraday as the script is runing to observe them turing ON and off.
 
-> Note: GPIO's being commanded may be recieving other commands and not work as intended if multiple programs are controlling a single GPIO (i.e. RED due to RF TX indication)
+## Tutorial Output Examples
+
+below is a screen-shot of the partial output of the tutorial script when run in a python interpreter (PyCharm). Be sure to look at the two LED's on Faraday as the script is running to observe them turing ON and OFF.
```
Comment (kb1lqc) [inline Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md]: Spelling "turning"

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md [hunk 12]
```diff
@@ -33,7 +37,7 @@ below is a screen-shot of the partial output of the tutorial script when run in
 
 ## Code - Toggle GPIO's (LEDs) Predefined Functions
 
-The Faraday command module object pre-defines many common actions such as turning ON/OFF the on-board LED's. Using `faraday_cmd.CommandLocalGPIOLED1On()` will command the LED #1 to an ON state and light up the LED. Alternatively `faraday_cmd.CommandLocalGPIOLED1Off()` turns the LED OFF.
+The Faraday command module object predefine's many common actions such as turning ON/OFF the on-board LED's. Using `faraday_cmd.CommandLocalGPIOLED1On()` will command the LED #1 to an ON state and light up the LED. Alternatively `faraday_cmd.CommandLocalGPIOLED1Off()` turns the LED OFF.
```
Comment (kb1lqc) [inline Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md]: predefine's is wrong should be predefines

Comment (kb1lqc) [reply Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md]: You also refer to `LED #1` here but `LED#1` earlier in the document

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md [hunk 13]
```diff
@@ -63,14 +67,14 @@ time.sleep(1)
 
 ## Code - Toggle GPIO's (LEDs) Bitmask
 
-The previous code used pre-defined functions for toggling GPIO and left LED #2 ON. Those predefined functions are actually just providing an abstraction of the raw command for GPIO bitmask control. The GPIO bitmask control is part of the Command Application and provides both ON/OFF control of many GPIO in a single command while also guarding potentially damaging GPIO from being accidently changed (i.e. removes ability to manually toggle the RF amplifier control GPIO).
+The predefined functions are actually just providing an abstraction of the `CommandLocalGPIO(self, p3_bitmask_on, p4_bitmask_on, p5_bitmask_on, p3_bitmask_off, p4_bitmask_off, p5_bitmask_off)` function that provides GPIO bitmask control of all controllable GPIO pins. This functionality is part of the Command Application and provides control of all GPIO in a single command while also guarding potentially damaging GPIO from being accidentally changed (i.e. removes ability to manually toggle the RF amplifier GPIO).
```
Comment (kb1lqc) [inline Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md]: > ...while also guarding potentially damaging GPIO from being accidentally changed
Weird wording

...while also guarding against changing GPIO values which could damaging a GPIO pin.

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md [hunk 14]
```diff
@@ -106,14 +110,15 @@ faraday_1.POST(local_device_callsign, local_device_node_id, faraday_1.CMD_UART_P
...
+A simple ECHO like command is provided that will send the supplied data payload back to the host computer. The ECHO command only accepts a single UART transport packet so it is limited in size to the 124 byte maximum transmissible unit of the current UART network stack.
 
-This command example is pretty self explaining and simply sends the string `"This will ECHO back on UART"` to the local Faraday device and retrieves, parses, and displays the ECHO'd packet. Note the padding bytes likely visible due to fixed length UART transport packet formatting appended to the original message. Also, this command accepts binary data values (not just ASCII string text)!
+This example sends the string `"This will ECHO back on UART"` to the local Faraday device and retrieves, parses, and displays the ECHO'd packet. Note the appended padding bytes visible due to fixed length UART transport packet.
```
Comment (kb1lqc) [inline Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md]: Better wording

...to the local Faraday device retrieving, parsing, and displaying the ECHO'd packet.

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md [hunk 15]
```diff
@@ -143,15 +150,18 @@ below is a screenshot of the partial output of the tutorial script when run in a
...
-#See Also
+![Output from the Commanding - Local Tutorial](Images/Echo_Exception.png "Output from the Commanding - Local Tutorial")
 
-* [Proxy Tool - FaradayIO](http://faraday-software.readthedocs.io/en/latest/faradayio.html)
-* [Proxy Tool - TelemetryParser](http://faraday-software.readthedocs.io/en/latest/telemetryparser.html)
-* Command Application
+#Bonus Exercise
+
+* Modify the the example script to remove the padding bytes and display only the original ECHO'd string.
+  * Can you support variably given any string length using only the packets sent and received ("Decoded BASE64 RAW Bytes")?
```
Comment (kb1lqc) [inline Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md]: What? Variably? Sentence does not make sense to me.

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md [hunk 16]
```diff
@@ -143,15 +150,18 @@ below is a screenshot of the partial output of the tutorial script when run in a
...
 
+# To-Do
```
Comment (kb1lqc) [inline Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md]: Why have todo's here?

**Review by kb1lqc — CHANGES_REQUESTED** [review]
Haven't yet run through using hardware but found enough spelling and wording stuff to have you get that working while I work on using this with hardware.

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py (hunk 3 above)
Comment (kb1lqd) [reply Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py]: Updated node ID to gitint() but need to verify working.

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py (hunk 1 above)
Comment (kb1lqd) [reply Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py]: Yep, completed.

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py (hunk 2 above)
Comment (kb1lqd) [reply Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py]: Well I agree our applications and larger tutorials should be upgraded to logging this is simply actively telling the user when the LED should be ON. Print is appropriate.

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py (hunk 4 above)
Comment (kb1lqd) [reply Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py]: Correct, the inline comment explains the intent to show the raw bitmask function as well.

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py (hunk 5 above)
Comment (kb1lqd) [reply Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py]: I moved the "sending..." print to above receiving code and after initializing the message.

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py (hunk 6 above)
Comment (kb1lqd) [reply Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py]: I updated to Type Error, I think this is correct? "except TypeError as e"

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/configuration-template.ini (hunk 7 above)
Comment (kb1lqd) [reply Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/configuration-template.ini]: Done

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md (hunk 8 above)
Comment (kb1lqd) [reply Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md]: Yeah I think I did the new file + .gitignore but not sure, can you check?

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md (hunk 9 above)
Comment (kb1lqd) [reply Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md]: Fixed

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md (hunk 10 above)
Comment (kb1lqd) [reply Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md]: Yep

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md (hunk 11 above)
Comment (kb1lqd) [reply Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md]: Fixed

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md (hunk 12 above)
Comment (kb1lqd) [reply Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md]: Where?

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md (hunk 13 above)
Comment (kb1lqd) [reply Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md]: Updated.

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md (hunk 14 above)
Comment (kb1lqd) [reply Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md]: Fixed

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md (hunk 16 above)
Comment (kb1lqd) [reply Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md]: Ok, I'll remove.

**Comment by kb1lqd** [comment]
@kb1lqc Please re-review and note my comments with questions/checks that I did things correctly. Mostly, did I update the .gitignore and rename the INI correctly.

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py (hunk 6 above)
Comment (kb1lqc) [reply Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py]: Well it depends.

If line 91 fails it could be an IndexError or KeyError
Like 92 and the print statements are likely just covered by a StandardError. The trick is you want to react accordingly and appropriately. While catching the exception can prevent program crashing you could insert valid but filler data or rollback to a safe state depending on the error and circumstances.

Instead of just having a KeyError for `b64_data = rx_echo_raw[0]['data']` there could be a common issue with Proxy where you can inform the user to check a specific input value or configuration.

File: Tutorials/Python_Developer_Tutorials/Readme.md [hunk 17]
```diff
@@ -19,7 +19,7 @@ Each tutorial series builds upon the prior to create a simple wireless transmiss
 
 ##Introduction To Faraday Proxy Programming
 
-* **[Commanding - Local](foundation/Commanding-Local/) (OUTDATED - UPDATE IN PROGRESS)**
+* **[Commanding - Local](foundation/Commanding-Local/)
```
Comment (kb1lqc) [inline Tutorials/Python_Developer_Tutorials/Readme.md]: Stray "**" here

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md [hunk 18]
```diff
@@ -1,29 +1,33 @@
...
+* Update `REPLACEME` from `CALLSIGN` to match the callsign of the Faraday unit **as assigned** in proxy
+* Update `REPLACEME` from `NODEID` to match the callsign node ID of the Faraday unit **as assigned** in proxy
+* Save the file as `command_local.ini`
 
-## Tutorial Output Examples
 
-below is a screen-shot of the partial output of the tutorial script when run in a python interpreter (PyCharm). Be sure to look at the two LED's on Faraday as the script is runing to observe them turing ON and off.
 
-> Note: GPIO's being commanded may be recieving other commands and not work as intended if multiple programs are controlling a single GPIO (i.e. RED due to RF TX indication)
+## Tutorial Output Examples
```
Comment (kb1lqc) [inline Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md]: You never actually tell the user how to run the program!

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py (hunk 6 above)
Comment (kb1lqc) [reply Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py]: First time I ran this program after starting Proxy:
image

Second time:
image

---
Yes if I restart Proxy I get the same error. Looking at Proxy the following error occurs the first time this program is run after starting Proxy:
image

What's up with this? I've confirmed this exists but also, you've sent commands to port 2 for LED's. I do not understand why only ECHO fails on the first try.

I printed out the rx_exho_raw and you are getting a JSON error message back. I reccommend using that error message in your exception code:
image

**Review by kb1lqc — CHANGES_REQUESTED** [review]
Some errors occurred when running. It appears that ECHO function acts differently than the LED functions causing the problems. I do not expect this:
image

Since LED commands were already sent several times prior, the port has been used!

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py [hunk 19]
```diff
@@ -1,87 +1,110 @@
...
-echo_decoded = faraday_1.DecodeRawPacket(b64_data)
-
-#Display information
-print "Original Message: ", originalmsg
-print "RAW Received BASE64 ECHO'd Message:", b64_data
-print "Decoded received ECHO'd Message:", echo_decoded #Note that ECHO sends back a fixed packed regardless. Should update to send back exact length.
-
-
+#  Retrive waiting data packet in UART Transport service number for the COMMAND application (Use GETWait() to block until ready or timeout
+rx_echo_raw = faraday_1.GETWait(local_device_callsign, local_device_node_id, faraday_1.CMD_UART_PORT,
```
Comment (kb1lqc) [inline Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py]: This should be checked for the "error" JSON response as shown below:
image

failing gracefully and printing the error message returned by proxy is what we should do. This is necessary in all applications but just noticed that we weren't using this functionality I built in to the API.

This is failing on first try because the GETWait() because you are trying to obtain data from the CMD_UART_PORT for the first time. All prior LED operations were a POST() to the port. This means that the buffer hasn't been created yet for this particular port on the RX side of things. This is Issue #60 creeping up.

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py [hunk 20]
```diff
@@ -1,87 +1,112 @@
...
+    echo_decoded = faraday_1.DecodeRawPacket(b64_data)
+
+    # Display information
+    print "\n**Receiving**\n"
+    print "Decoded received ECHO'd Message:", echo_decoded  # Note that ECHO sends back a fixed packed regardless. Should update to send back exact length.
+    print "\nRAW Received BASE64 ECHO'd Message:", b64_data
+    print "\nDecoded BASE64 RAW Bytes:", repr(echo_decoded)
+
+except Exception as e:
+    print "Failed ECHO due to type error:", e
```
Comment (kb1lqc) [inline Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Proxy-Local-Commands-Tutorial_.py]: Not really a type error. Would be MOST useful to print out the error JSON message here.

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md [hunk 21]
```diff
@@ -143,15 +150,13 @@ below is a screenshot of the partial output of the tutorial script when run in a
 
 The padding bytes are clearly visible appended to the end of the returned ECHO'd string. 
 
-#Bonus Excersize
+### ECHO Exception - Failure
 
-* Modify the the example script to remove the padding bytes and display only the original ECHO'd string.
-* Can you do this variably given any string length using only the packets sent and received?
-  * *Hint: [Application layer](https://en.wikipedia.org/wiki/OSI_model) packet [encapsulation](https://en.wikipedia.org/wiki/Encapsulation_(networking))*
+There is a [known bug](https://github.com/FaradayRF/Faraday-Software/issues/50) that may cause the first packet from Proxy to not properly be retrieved and this has caused ECHO to fail. The workaround to this until it is patched is to simply re-run the tutorial script **without** restarting proxy.
```
Comment (kb1lqc) [inline Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md]: Ahh I missed that you noted this already. It's issue #60 not 50

**Comment by kb1lqd** [comment]
@kb1lqc I updated per comments again:

* Added section to instruct how to run tutorial script
* Moved known bug description/info to top of readme in tutorial script running section
* Changed general exception to "KeyError" as expected by known bug error. I also added a print statement to describe the bug/workaround

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md [hunk 22]
```diff
@@ -63,14 +79,14 @@ time.sleep(1)
 
 ## Code - Toggle GPIO's (LEDs) Bitmask
 
-The previous code used pre-defined functions for toggling GPIO and left LED #2 ON. Those predefined functions are actually just providing an abstraction of the raw command for GPIO bitmask control. The GPIO bitmask control is part of the Command Application and provides both ON/OFF control of many GPIO in a single command while also guarding potentially damaging GPIO from being accidently changed (i.e. removes ability to manually toggle the RF amplifier control GPIO).
+The predefined functions are actually just providing an abstraction of the `CommandLocalGPIO(self, p3_bitmask_on, p4_bitmask_on, p5_bitmask_on, p3_bitmask_off, p4_bitmask_off, p5_bitmask_off)` function that provides GPIO bitmask control of all controllable GPIO pins. This functionality is part of the Command Application and provides control of all GPIO in a single command while also guarding against changing GPIO values which could damage Faraday (i.e. removes ability to manually toggle the RF amplifier GPIO).
```
Comment (kb1lqc) [inline Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md:82]: Since when did we get a "Command Application"?

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md [hunk 23]
```diff
@@ -63,14 +79,14 @@ time.sleep(1)
...
 
 The Python module function `faraday_cmd.CommandLocalGPIO()` contains 6 function arguments:
 
 * 3 function arguments for commanding GPIO ports 3, 4, 5 pins ON
 * 3 function arguments for commanding GPIO ports 3, 4, 5 pins OFF
 
-If any bits are HIGH (1) in the bitmask bytes then that respective pin will be toggled ON (HIGH), if a bit is left LOW (0) then no action will take place and the bit will remain in it's current HIGH/LOW state.
+If any bits are HIGH (1) in the bitmask bytes then the respective ON/OFF action will be applied to that bitmask's pin, if a bit is left LOW (0) then no action will take place and the pin will remain in it's current HIGH/LOW state. Commanding both and HIGH and LOW for the same port/pin is not allowed.
```
Comment (kb1lqc) [inline Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md:89]: what happens when you command both ON and OFF at the same time?

**Comment by kb1lqd** [comment]
The python module function won’t allow it and will fail and should return False IIRC.

File: Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md [hunk 24]
```diff
@@ -143,15 +162,7 @@ below is a screenshot of the partial output of the tutorial script when run in a
...
-* Can you do this variably given any string length using only the packets sent and received?
-  * *Hint: [Application layer](https://en.wikipedia.org/wiki/OSI_model) packet [encapsulation](https://en.wikipedia.org/wiki/Encapsulation_(networking))*
-
-#See Also
-
-* [Proxy Tool - FaradayIO](http://faraday-software.readthedocs.io/en/latest/faradayio.html)
-* [Proxy Tool - TelemetryParser](http://faraday-software.readthedocs.io/en/latest/telemetryparser.html)
-* Command Application
-
+  * Can you do this with variable length ECHO data?
```
Comment (kb1lqc) [inline Tutorials/Python_Developer_Tutorials/foundation/Commanding-Local/Readme.md:168]: Your example doesn't show any extra text on the echo'd string. I don't understand.

**Review by kb1lqc — APPROVED** [review]
Items I think need to be fixed in new Issue Ticket:
- Wording issues that make text hard to read
- Example code has long comments that make text hard to follow/scroll
- Formatting discrepancies
- Spelling errors still present
- Example output doesn't match code

I think we need to get this into the master and fix these issues on a new PR.
